### PR TITLE
feat: serve sync.proto on port 8015

### PR DESCRIPTION
**Pull request by Kavindu-Dodan**
## This PR

Introduce flag sync capability to flagd as discussed at [1] and fixes https://github.com/open-feature/flagd/issues/1230 .

**What's changed ?**

The change included with this PR introduces the gRPC sync contract [2]. This allows flagd to expose its store to in-process provider which consumes gRPC flag stream. 

image

**How to use ?**

flagd will start sync service on port `8015`. You can alter the default sync port by providing desired port to the startup flag `syncPort` (`--sync-port=8686`) 

**Implementation details**

- Sync service startup is delayed to allow configured flag sources to complete their initial loading (dealy is 5 seconds)
- Sync request's `selector` can be used to specify the specific flag source. If unset, all flags will be sent
- A new stream response will be created whenever there are updates from flag sources 

[1] - https://github.com/open-feature/flagd/discussions/1153 
[2] - https://github.com/open-feature/flagd-schemas/blob/main/protobuf/flagd/sync/v1/sync.proto

### CODE REVIEW

File: flagd/cmd/start.go
Comment (bacherfl) [inline flagd/cmd/start.go:29]: should we change this to `sync-enabled` and `sync-port`, to stay in line with the other flags which are all in kebap-case?

File: flagd/pkg/runtime/from_config.go
Comment (bacherfl) [inline flagd/pkg/runtime/from_config.go:15]: nitpick, to use the recommended [naming scheme](https://go.dev/blog/package-names) : 
```suggestion
	flagsync "github.com/open-feature/flagd/flagd/pkg/service/flag-sync"
```

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (bacherfl) [inline flagd/pkg/service/flag-sync/sync-multiplexer.go:138]:
```suggestion
func (r *Multiplexer) GetAllFlags(source string) (string, error) {
```

File: flagd/cmd/start.go
Comment (Kavindu-Dodan) [reply flagd/cmd/start.go:29]: Good point, will change to kebab-case

Comment (Kavindu-Dodan) [reply flagd/cmd/start.go]: done this the commit https://github.com/open-feature/flagd/pull/1237/commits/d7c5798048864dcc504fffafbdebbf61fbdf1938

File: flagd/pkg/runtime/from_config.go
Comment (Kavindu-Dodan) [reply flagd/pkg/runtime/from_config.go]: agree, done with the commit https://github.com/open-feature/flagd/pull/1237/commits/d7c5798048864dcc504fffafbdebbf61fbdf1938

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync-multiplexer.go]: commit https://github.com/open-feature/flagd/pull/1237/commits/d7c5798048864dcc504fffafbdebbf61fbdf1938 fixes this

File: flagd/cmd/start.go
Comment (beeme1mr) [inline flagd/cmd/start.go:70]:
```suggestion
	flags.BoolP(syncEnabledFlagName, "e", false, "Enables the gRPC sync service. "+
```

Comment (beeme1mr) [reply flagd/cmd/start.go:70]: I could go either way on this one. I was trying to make it sound similar to the other description.

**Review by beeme1mr — APPROVED** [review]

File: flagd/cmd/start.go
Comment (Kavindu-Dodan) [reply flagd/cmd/start.go]: I like the suggestion, fixed with commit https://github.com/open-feature/flagd/pull/1237/commits/bf818c7a163d4b019e408fce9653bcc4a72aa1b4 as we need to update documentation as well

File: flagd/pkg/runtime/runtime.go
Comment (thisthat) [inline flagd/pkg/runtime/runtime.go:107]: why this timeout? I don't think we can guarantee that all sync fetch their flags in 5s 🤔

Comment (bacherfl) [reply flagd/pkg/runtime/runtime.go:107]: do we even need to wait for the initial flag retrieval to be finished here? Connecting to the flag sync endpoint will establish a stream over which any future updates will be communicated anyway, so if the initial sync is not finished at the time a client calls the `Sync` endpoint, the fetched configuration will be passed to any connected client anyway. However, please correct me if there is something I am missing / not considering here

File: docs/architecture.md
Comment (toddbaert) [inline docs/architecture.md:40]:
```suggestion
This pattern requires an in-process implementation of the flagd evaluation engine but has the benefit of no I/O overhead for flag evaluations, since no inter-process communication is required.
```

File: docs/reference/grpc-sync-service.md
Comment (toddbaert) [inline docs/reference/grpc-sync-service.md:7]:
```suggestion
flagd can expose a gRPC sync service, allowing in-process providers to obtain their flag definitions.
```

File: docs/reference/grpc-sync-service.md
Comment (toddbaert) [inline docs/reference/grpc-sync-service.md:25]:
```suggestion
By default, the gRPC stream exposes all the flag configurations merged consistent with the flag's standard merge strategy.
```

File: docs/reference/grpc-sync-service.md
Comment (toddbaert) [inline docs/reference/grpc-sync-service.md:28]:
```suggestion
If you specify a `selector` in the gRPC sync request, the gRPC service will attempt match the provided selector value to a source, and stream just the flags identified in that source.
```

File: docs/reference/grpc-sync-service.md
Comment (toddbaert) [inline docs/reference/grpc-sync-service.md:29]: Does the selector have to match the CLI arg exactly? If so, we should probably make it `file://myFlags.json` for clarity.

File: flagd/cmd/start.go
Comment (toddbaert) [inline flagd/cmd/start.go:29]: Not sure, but we could _probably_ just remove this flag and always serve this. If you don't want to use it, you can simply not expose it on the network. Interested in other opinions on this one.

File: flagd/cmd/start.go
Comment (toddbaert) [inline flagd/cmd/start.go:70]: See [this](https://github.com/open-feature/flagd/pull/1237/files#r1517882344).

File: flagd/pkg/runtime/runtime.go
Comment (toddbaert) [reply flagd/pkg/runtime/runtime.go:107]: > if the initial sync is not finished at the time a client calls the Sync endpoint, the fetched configuration will be passed to any connected client anyway.

This is true, but there's a potential downside here, since the initial sync response is what providers use to signal their READY state. If they get this message, they will assume flags are ready to be evaluated, though in reality they are not since flagd hasn't reconciled it's internal state yet from all sources and only a partial (or possibly empty) set is returned.

I think the best solution would be to delay any sync responses until all sources have resolved, OR some timeout is reached (in which case an error is logged).

Comment (Kavindu-Dodan) [reply flagd/pkg/runtime/runtime.go:107]: This is how I looked at this, 

When there are multiple syncs (ex:- file sync + http sync + ...), each initial sync may complete with some delay. Consider the following mermaid diagram, 

```mermaid
gantt
    title Data Syncs & Service start
    dateFormat  s
    axisFormat %S

    section Data Syncs
    Sync 1          :a1, 1, 2s
    Sync 2          :a2, 1, 3s
    Sync 3          :a3, 2, 4s
    
    section Sync Service

    Service Start & running    :b2, 6, 10s
```

If we start the sync service without some delay, then sync subscriptions will receive multiple syncs till all data syncs stabilize with the initial data fetch. But if we add a start delay, at least we can attempt to stream stable data stream. 


 > I don't think we can guarantee that all sync fetch their flags in 5s  @thisthat 
 
 💯 , this is true and a 5s delay doesn't guarantee that all syncs are complete. We could optimize this with some tracking functionality but I did not invest time on this optimization. The same is true for re-syncs (especially after a delete). I think current implementation is good enough till we see the need to improve the logic
 
 > Connecting to the flag sync endpoint will establish a stream over which any future updates will be communicated anyway @bacherfl 
 
 Correct, this delay is only there for initial data sync stabilization. Like I explained above, not the optimal solution but I thought it's good to add this.
 
 And once service start, it's business as usual. Any data sync that doesn't trigger a resync will trigger a new sync stream update. 
 
 > Why not at resync ? 
 
 A resync means data store is not stable (a delete has occured). So there will be new data fetchings. Here too we can optmize to emit streams only when all resyncs are done. But I did not do this optimization either :)

Comment (toddbaert) [reply flagd/pkg/runtime/runtime.go:107]: > I think current implementation is good enough till we see the need to improve the logic

I also agree with this, but I think we should add an issue mentioning it. Something that basically adds a configurable timeout, but otherwise waits for every sync to initialize before responding the first time. Happy to consider this convo resolved if we define such an issue.

Comment (Kavindu-Dodan) [reply flagd/pkg/runtime/runtime.go:107]: >  since the initial sync response is what providers use to signal their READY state @toddbaert 

Yes, this is a really good point. And I will spend some time to see if we can optimize this 🤔 Or at least create a follow-up to improve this further

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (toddbaert) [inline flagd/pkg/service/flag-sync/sync_service.go:70]: I dunno if it's something we should add here, but I think there's some missing context/cancellation handling somewhere.

If I send a `SIGINT` to flagd WITHOUT the `-e` flag, it shuts down fine within a couple seconds. If I send a `SIGINT` WITH the `-e` flag, flagd never shuts down.

File: docs/reference/grpc-sync-service.md
Comment (Kavindu-Dodan) [reply docs/reference/grpc-sync-service.md:29]: Good question, I had to double check on this. 

No, this doesn't have to match CLI arg of URI format. It's only the source. For example if `--uri=file://myFlags.json` is used, then selector should be `myFlags.json`.

**Review by toddbaert — CHANGES_REQUESTED** [review]
Looks really good, looks very clean for how much new functionality is here, but I guess part of that is because you did that nice refactor in advance, good job.

I tested this locally, and besides [this](https://github.com/open-feature/flagd/pull/1237#discussion_r1517923018) and [this](https://github.com/open-feature/flagd/pull/1237#issuecomment-1986008374), things work as expected. I tested both the `Sync` and the `FetchAllFlags` methods. I think I can approve once these are addressed.

File: flagd/cmd/start.go
Comment (Kavindu-Dodan) [reply flagd/cmd/start.go:29]: I added this service as "disabled by default" so that users won't be affected by a new service when we do the next release. This can happen if a K8s deployment uses the latest tag or by mistake if someone forgets to check release notes.

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync_service.go]: Good observation. I checked the shutdown logic and improved it to respect the SIGTERM. Updated the PR with commit https://github.com/open-feature/flagd/pull/1237/commits/711a21b916f8f4e65aa8c30fcc58b8df8dcc3293

**Comment by Kavindu-Dodan** [comment]
> This seems a bit brittle and confusing, but I think it's out of scope of this PR. I think we should just document it for now, and perhaps add a new PR that allows users to define an friendly label for each source in the JSON config mode. @toddbaert 

Agree. The issue is how we handle `--uri` or `-f` , we convert the parameter to a `SourceConfig` [1] at startup. And then when updates are emitted from data syncs (ex:- file, k8s), the payload only contains the URI, which is merely the source without any prefix. 

I Will update the documentation highlighting this detail. 


[1] - https://github.com/open-feature/flagd/blob/main/core/pkg/sync/builder/utils.go#L36

File: flagd/cmd/start.go
Comment (toddbaert) [reply flagd/cmd/start.go:29]: > I added this service as "disabled by default" so that users won't be affected by a new service when we do the next release. This can happen if a K8s deployment uses the latest tag or by mistake if someone forgets to check release notes.

That makes sense, however, since this is on a different port, I would be surprised if there was any severe networking or security implications since no forwarding/exposure would be occurring on this port from the wider infra. I think just mentioning this explicitly on the docs and title should be enough to assuage that concern.

I don't feel strongly about this point, but I think it's a bit odd having to opt into this considering we don't require users to opt into the evaluation.proto, and they are both equally valid to use.

My suggestion would be to remove this and change the title to `feat: serve sync.proto on port 8015` which will make things very clear.

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (toddbaert) [reply flagd/pkg/service/flag-sync/sync_service.go]: Tested and works great now.

**Review by toddbaert — APPROVED** [review]
SIGINT and docs look good now. Thanks a lot @Kavindu-Dodan

File: flagd/pkg/runtime/runtime.go
Comment (Kavindu-Dodan) [reply flagd/pkg/runtime/runtime.go]: After considering several options, I implemented a simple solution. Please see commit https://github.com/open-feature/flagd/pull/1237/commits/4e4e4df988345c8206ca837112dce14daf50f224

- I have transferred startup delay as well as emit condition to sync_service 

This means runtime no longer track resync condition nor it delay service startup. This responsiblity is at sync_service

- A simple count down latch `syncTracker` tracks flag sync updates 
- Server startup is delayed till either `syncTracker` emit a rediness OR a timeout
- `syncTracker` emit a ready when all known sync sources emit at least one sync payload 

I think this is good enough solution till we see more improvements :)

**Comment by Kavindu-Dodan** [comment]
> SIGINT and docs look good now. Thanks a lot @Kavindu-Dodan

Thank you for detailed reviews 🤝

File: flagd/cmd/start.go
Comment (Kavindu-Dodan) [reply flagd/cmd/start.go:29]: I will wait for @beeme1mr , @thisthat and @bacherfl feedback and do this change if all of is agree :)

Comment (bacherfl) [reply flagd/cmd/start.go:29]: I agree, I think if this new endpoint is highlighted explicitly in the release notes, the flag should not be required

File: flagd/pkg/runtime/runtime.go
Comment (bacherfl) [reply flagd/pkg/runtime/runtime.go]: thanks for adding this @Kavindu-Dodan, I think that's a good solution :)

File: flagd/cmd/start.go
Comment (thisthat) [reply flagd/cmd/start.go:29]: I like the idea of hiding the new functionalities behind a "release feature flag" :p But I also agree with the points of @toddbaert and I don't think we really need this

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync_service.go:140]: maybe `trackAndRemove`

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync_service.go:136]: since we don't expose this function, why we need this getter? 🤔

File: flagd/pkg/runtime/runtime.go
Comment (Kavindu-Dodan) [reply flagd/pkg/runtime/runtime.go]: Done, removed the enabling flag with commit https://github.com/open-feature/flagd/pull/1237/commits/1c330ff9dcfcf9fdab4a3a6031b29c50ff3e53f3

File: flagd/cmd/start.go
Comment (Kavindu-Dodan) [reply flagd/cmd/start.go]: Done, I have removed the flag with commit https://github.com/open-feature/flagd/pull/1237/commits/1c330ff9dcfcf9fdab4a3a6031b29c50ff3e53f3

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync_service.go:80]: I am not a native speaker but this sounds more readable 🤔 
```suggestion
			s.logger.Warn("finish waiting for all sync sources to complete their initial sync. " +
```

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync-multiplexer.go:62]: do we document this behavior? I was under the impression that if we don't provide any source then we don't return any flags

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync-multiplexer.go:53]: do we really need to lock here?

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync-multiplexer.go:152]:
```suggestion
	r.mu.RLock()
```

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (thisthat) [inline flagd/pkg/service/flag-sync/sync-multiplexer.go:139]: I think we should read-lock here as well

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync-multiplexer.go:62]: This behaviour is agreed and defined through the issue itself :)

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync_service.go:136]: Given the struct is in the same module, yes it's possible to access the internal channel. However, as a best practice, I like to avoid directly using internals and rely on a getter instead.

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync_service.go]: done, updated :)

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync_service.go]: I tried to improve this log. Here the meaning should include a timeout as this is a timeout waiting for all sync sources to complete.

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync-multiplexer.go]: done :)

File: flagd/pkg/service/flag-sync/sync-multiplexer.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync-multiplexer.go:142]: This is done :)

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (thisthat) [reply flagd/pkg/service/flag-sync/sync_service.go:136]: "done" feels like closing the `syncTracker`, maybe we could call it `getDone()` 🤔

**Review by thisthat — APPROVED** [review]

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (toddbaert) [reply flagd/pkg/service/flag-sync/sync_service.go:136]: It's a nit but I kinda agree.

**Comment by Kavindu-Dodan** [comment]
Merging with all the approvals. Thank you @thisthat @toddbaert @bacherfl @beeme1mr for reviews 🤝

File: flagd/pkg/service/flag-sync/sync_service.go
Comment (Kavindu-Dodan) [reply flagd/pkg/service/flag-sync/sync_service.go]: This is done :)

**Pull request merged**
